polygeist/lib/polygeist/Passes/AffineReduction.cpp

```diff
@@ -1,264 +1,313 @@
-        }
-        // require a single store within the current for. The load must dominate
-        // the single store. There must be no other stores in the current for.
-        if ((candidateStores.size() == 1) &&
-            checkDominance(load.getOperation(), candidateStores[0].getOperation()) &&
-            otherStores.size() == 0 /*
-            checkDominance(candidateStores[0].getOperation(), otherStores)*/) {
-          candidateOpsInFor.push_back(std::make_pair(
-              load.getOperation(), candidateStores[0].getOperation()));
-          loadsInFor.push_back(otherLoads);
-        }
+    AffineLoadOp getLoad() const { return Load; }
+    AffineStoreOp getStore() const { return Store; }
+
+  private:
+    AffineLoadOp Load;
+    AffineStoreOp Store;
+  };
+
+  /// Collect candidate reduction operations in the given loop \p ForOp.
+  /// Candidate reductions are stored into \p CandidateOps, and affine load
+  /// operations with the same base operand and subscript indices as the
+  /// candidate load are stored into \p LoadsInLoops.
+  WalkResult
+  collectCandidates(AffineForOp ForOp,
+                    SmallVectorImpl<CandidateReductionOp> &CandidateOps,
+                    SmallVector<Operation *> &LoadsInLoop) const {
+    assert(CandidateOps.empty() && "Expecting an empty vector");
+    assert(LoadsInLoop.empty() && "Expecting an empty vector");
+
+    WalkResult Result = ForOp.getBody()->walk([&](Operation *Op) {
+      // We are only interested in affine load operations.
+      if (!isa<AffineLoadOp>(Op))
+        return WalkResult::advance();
+
+      // Skip the load if any of its subscript indices are the loop induction
+      // variable (i.e. the load is not loop invariant).
+      auto Load = cast<AffineLoadOp>(Op);
+      SmallVector<Value, 4> Indices(Load.getIndices());
+      if (!hasAllDimsReduced(Indices, ForOp.getInductionVar()))
+        return WalkResult::advance();
+
+      // Locate possible compatible stores (stores that have the same base
+      // operand and subscript indices as the load), and collect all other loads
+      // that have the same subscript and base symbol.
+      SmallVector<AffineStoreOp> CandidateStores;
+      SmallVector<Operation *> OtherStores;
+      for (Operation *User : Load.getMemRef().getUsers()) {
+        TypeSwitch<Operation *>(User)
+            .Case<AffineLoadOp>([&](auto OtherLoad) {
+              if (areCompatible(Load, OtherLoad) && Load != OtherLoad &&
+                  isNestedInForOp(OtherLoad.getOperation(), ForOp))
+                LoadsInLoop.push_back(OtherLoad);
+            })
+            .Case<AffineStoreOp>([&](auto Store) {
+              if (areInSameAffineFor(Load, Store, ForOp) &&
+                  areCompatible(Load, Store))
+                CandidateStores.push_back(Store);
+              else if (areCompatible(Load, Store) &&
+                       isNestedInForOp(Store.getOperation(), ForOp))
+                OtherStores.push_back(Store);
```

Comment:
We are inside a `TypeSwitch` that does not return anything, so we cannot return  `WalkResult::interrupt()`. I could make the TypeSwitch return a bool and then check whether the walk should continue or not. There isn't much advantage in doing it that way though and later, if we want to extend the transformation, it might be useful to collect other compatible stores.